compiler/src/dotty/tools/dotc/transform/Pickler.scala

```diff
@@ -45,6 +46,42 @@ class Pickler extends Phase {
     clss.filterNot(companionModuleClasses.contains)
   }
 
+  class PickleCompleter:
+    private var pickled: Array[Byte] = null
+
+    def complete(pickler: TastyPickler, cls: ClassSymbol, tree: Tree)(using Context): Unit =
+      val treePkl = pickler.treePkl
+      if tree.span.exists then
+        new PositionPickler(pickler, treePkl.buf.addrOfTree).picklePositions(tree :: Nil)
+
+      if !ctx.settings.YdropComments.value && false then
+        new CommentPickler(pickler, treePkl.buf.addrOfTree).pickleComment(tree)
+
+      val pickled = pickler.assembleParts()
+
+      def rawBytes = // not needed right now, but useful to print raw format.
+        pickled.iterator.grouped(10).toList.zipWithIndex.map {
+          case (row, i) => s"${i}0: ${row.mkString(" ")}"
+        }
+
+      synchronized {
+        this.pickled = pickled
+        // println(i"rawBytes = \n$rawBytes%\n%") // DEBUG
+        if pickling ne noPrinter then
+          pickling.synchronized {
+            println(i"**** pickled info of $cls")
+            println(new TastyPrinter(pickled).printContents())
+          }
+        notifyAll()
+      }
+    end complete
+
+    def bytes: Array[Byte] = synchronized {
+      if pickled == null then wait()
```

Comment:
This should be a while to deal with spurious wakeups:
```suggestion
      while pickled == null then wait()
```

Reply:
I based it on a Future in the end.